File: Python/Minimum_swaps_for_balanced_brackets.py
"""
Idea:
keeping count of left & right brackets, and adding value of imbalance to ans
"""


# User function Template for python3
class Solution:
    def minimumNumberOfSwaps(self, S):
        countLeft = countRight = ans = imbalance = 0
        for i in range(len(S)):
            if S[i] == '[':
                countLeft += 1
                if imbalance > 0:
                    # If there is an imbalance, we need to swap brackets
                    ans += imbalance
                    imbalance -= 1
            elif S[i] == ']':
                countRight += 1
                imbalance = countRight - countLeft
        return ans


# code here


# {
# Driver Code Starts
# Initial Template for Python 3
if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        S = str(input())
        ob = Solution()
        print(ob.minimumNumberOfSwaps(S))
# } Driver Code Ends

File: C++/Minimum_swaps_for_balanced_brackets.cpp
/*
Idea:
keeping count of left & right brackets, and adding value of imbalance to ans
*/


#include<bits/stdc++.h> 
using namespace std;


class Solution {
public:
    int minimumNumberOfSwaps(string S)
    {
        int countLeft = 0, countRight = 0;
        int swap = 0, imbalance = 0;
        for (int i = 0; i < S.length(); i++)
        {
            if (S[i] == '[')
            {
                countLeft++;

                if (imbalance > 0)
                    // If there is an imbalance, we need to swap brackets
                    swap += imbalance--;
            }
            else if (S[i] == ']')
            {
                countRight++;
                imbalance = (countRight - countLeft);
            }
        }
        return swap;


    }
};


int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        string S;
        cin >> S;
        Solution ob;
        cout << ob.minimumNumberOfSwaps(S) << endl;
    }
    return 0;
}
